README, tools/dm: update wording (#893)
pingcap/docs: tools/dm/shard-merge.md

# Merge and Replicate Data from Sharded Tables

This document introduces the sharding support feature provided by Data Migration (DM). This feature allows you to merge and replicate the data of tables with the same table schema in the upstream MySQL or MariaDB instances into one same table in the downstream TiDB. It supports not only replicating the upstream DML statements, but also coordinating to replicate the table schema change using DDL statements in multiple upstream sharded tables.

> > **Note:** To merge and replicate data from the sharded tables, you need to set `is-sharding` to "true" in the task configuration file.

Reviewer: it's right

> > **Note:** To merge and replicate data from the sharded tables, you need to set `is-sharding` to "true" in the task configuration file.

Reviewer 2: @GregoryIan Do you mean the version provided by @CaitinChen is right? Please note that when the user edits the config file, this item might not exist.

> > **Note:** To merge and replicate data from the sharded tables, you need to set `is-sharding` to "true" in the task configuration file.

Author: @GregoryIan Please share your idea.

> > **Note:** To merge and replicate data from the sharded tables, you need to set `is-sharding` to "true" in the task configuration file.

Reviewer: `is-sharding: true` is better 🤔

> > **Note:** To merge and replicate data from the sharded tables, you need to set `is-sharding` to "true" in the task configuration file.

Author: Thanks.

- If you need to change the [table routing](../../tools/dm/data-synchronization-features.md#table-routing) rule, you have to wait for the replication of all sharding DDL statements to complete.
    - During the replication of sharding DDL statements, an error is reported if you use `dmctl` to change `router-rules`.
- If you need to `CREATE` a new table to a sharding group where DDL statements are being executed, you have to make sure that the table schema is the same as the newly modified table schema.
    - For example, both the original `table_1` and `table_2` have two columns (a, b) initially, and have three columns (a, b, c) after the sharding DDL operation, so after the replication the newly created table should also have three columns (a, b, c).
- Because the DM-worker that has received the DDL statements will pause the task to wait for other DM-workers to receive their DDL statements, the delay of data replication will be increased.

### Background

> Currently, DM uses the binlog in the `ROW` format to perform the replication task. The binlog does not contain the table schema information. When you use the `ROW` binlog to replicate data, if you have not replicated multiple upstream tables into the same downstream table, then there only exist DDL operations of one upstream table that can update the table schema of the downstream table. During the replication process, the DML statements can be constructed accordingly with the column values and the downstream table schema.  

Reviewer 2: Why remove this sentence?

**The `ROW` binlog can be considered to have the nature of self-description.**

Author: I will add it back.

### Principles

This section shows how DM replicates DDL statements in the process of merging sharded tables based on the above example in the [background](#background) section.

![shard-ddl-flow](../../media/shard-ddl-flow.png)

In this example, `DM-worker-1` replicates the data from MySQL instance 1 and `DM-worker-2` replicates the data from MySQL instance 2. `DM-master` coordinates the DDL replication among multiple DM-workers. Starting from `DM-worker-1` receiving the DDL statements, the DDL replication process is simplified as follows:

1. `DM-worker-1` receives the DDL statement from MySQL instance 1 at `t1`, pauses the data replication of the corresponding DDL and DML statements, and sends the DDL information to `DM-master`.

> 2. `DM-master` decides whether to coordinate the replication of this DDL statement based on the received DDL information, creates a lock for this DDL statement, sends the DDL lock information back to `DM-worker-1` and marks `DM-worker-1` as the owner of this lock at the same time.

Reviewer 2: It seems that using `whether` in this example is inappropriate. Same for other similar "whether" below.

Reviewer 3: Agree. DM-master always coordinates the replication if needed.

Reviewer 2: @CaitinChen So please update. Thanks!

Because data comes from the same MySQL instance, all the data is obtained from the same binlog stream. In this case, the time sequence is as follows:

1. The Syncer unit in DM-worker receives the DML statements of `schema V1` from both sharded tables when the replication begins.
2. At `t1`, the Syncer unit in DM-worker receives the DDL statements of `table_1`.
3. From `t2` to `t3`, the received data includes the DML statements of `schema V2` from `table_1` and the DML statements of `schema V1` from `table_2`.
4. At `t3`, the Syncer unit in DM-worker receives the DDL statements of `table_2`.
5. From `t4` on, the Syncer unit in DM-worker receives the DML statements of `schema V2` from both tables.

> If the DDL statements are not processed particularly during the data synchronization, when the DDL statement of `table_1` is replicated to the downstream and changes the downstream table schema, the DML statement of `schema V1` from `table_2` cannot be replicated successfully. Therefore, within a single DM-worker, a logical sharding group similar to that within `DM-master` is created, except that members of this group are different sharded tables in the same upstream MySQL instance.

Reviewer 2: ```suggestion
If you do not handle the DDL statements particularly during the data replication, when the DDL statement of `table_1` is replicated to the downstream and changes the downstream table schema, the DML statement of `schema V1` from `table_2` cannot be replicated successfully. Therefore, within a single DM-worker, a logical sharding group similar to that within `DM-master` is created, except that members of this group are different sharded tables in the same upstream MySQL instance.
```
